### ISSUE TITLE
Motion-Specific 3D VAE

### ISSUE THREAD

**Issue by antragoudaras** [alibaba/Tora#39]
First of all, congratulations on this very interesting and powerful work! The open-sourcing of Tora is a fantastic contribution to the community.

I am currently exploring the possibility of using your **Trajectory Extractor (TE)** module independently. Yet the goal is the same as in your TE module which takes a trajectory, represented as a `(L, H, W, 2)` map, convert it to a `g_vis` flow map (https://github.com/alibaba/Tora/blob/main/sat/utils/flow_utils.py#L120-L143), and use the motion-specific VAE to obtain the spacetime motion latent representation `g_m`.

In studying the paper and the codebase, I've encountered a few points that I would be very grateful if you could clarify. My questions primarily revolve around the **Motion-Specific 3D VAE** used within the Trajectory Extractor.

---

#### 1. Training of the Motion-Specific 3D VAE

The paper's Appendix (Section 11, "Motion VAE Training") details a specific fine-tuning process for this VAE:

> "We fine-tune the motion 3D VAE for 200,000 iterations with a batch size of 1... Our training data is sourced from a combination of datasets annotated with optical flow information [6, 21, 22, 33]."

And in Section 4.1:

> "The 3D VAE is initially trained on datasets [6, 21, 22, 33] for optical flow estimation and then **frozen during Tora training**."

This suggests a distinct, two-stage process where a specialized VAE is first trained on optical flow datasets and then used (frozen) within the main Tora framework.

However, in the codebase and in the discussion in GitHub issue #28 , it seems the TE module (including its VAE?) is fine-tuned during Stage 1 of the main Tora training, using the OpenSora/CogVideoX backbone.

**Could you please clarify this apparent discrepancy?**
*   Was a separate Motion-Specific 3D VAE pre-trained for 200k steps on datasets like KITTI-Flow, [Learning Human Optical Flow](https://humanflow.is.tue.mpg.de/index.html) and etc., as described in the appendix?
*   If so, are the weights for this specialized VAE available? Or is the correct approach to fine-tune the standard video VAE on motion maps as part of the main model's training?

---

#### 2. VAE Sharing and Architecture

This leads to my second question about the VAE's architecture and weights.

*   **Is the 3D VAE used in the Trajectory Extractor the *same*/*shared* VAE (i.e., shared weights) as the one used to encode the main video frames(following the [LDM approach](https://arxiv.org/pdf/2112.10752) )?** Or are they two separate VAEs one dedicated to video content and one for motion maps? 

*   The checkpoint structure (see bellow) contains a `vae/3d-vae.pt` file. Is this:
    a) The standard CogVideoX VAE's trained model?
    b) The weights for the **fine-tuned Motion-Specific VAE** described in the paper?
    c) Something else?

Understanding whether the motion-specific VAE is part of the main `mp_rank_00_model_states.pt` or the `vae/3d-vae.pt` checkpoint, that would also be very helpful to know for my project. Also is the `CogVideoX-5b-sat` folder which is missing from your HuggingFace referring to CogVideo-5B-sat, T2V 1.0 versions, the one before 1.5 or something else?

```
Tora
└── sat
    └── ckpts
        ├── t5-v1_1-xxl
        │   ├── model-00001-of-00002.safetensors
        │   └── ...
        ├── vae
        │   └── 3d-vae.pt
        ├── tora
        │   ├── i2v
        │   │   └── mp_rank_00_model_states.pt
        │   └── t2v
        │       └── mp_rank_00_model_states.pt
        └── CogVideoX-5b-sat # for training stage 1
            └── mp_rank_00_model_states.pt
```
---

#### 3. Role of `VAETemporalEmulator` and `VAESpatialEmulator`

In the `traj_module.py` file , there are intriguing classes named [`VAETemporalEmulator`](https://github.com/alibaba/Tora/blob/main/sat/sgm/modules/traj_module.py#L132-L153) and [`VAESpatialEmulator`](https://github.com/alibaba/Tora/blob/main/sat/sgm/modules/traj_module.py#L109-L132). These seem designed to mimic parts of the VAE's functionality but are either not used or commented out in the final `TrajExtractor` implementation.

*   Could you shed some light on the purpose of these emulator classes? Were they part of an earlier experimental design, and what was the reason for moving to the *specific* 3D Motion VAE approach instead?

---

#### 4. OpenSora vs. CogVideoX Discrepancy

Finally, the paper frequently describes the approach as being based on the **OpenSora** framework, but the provided codebase, model weights (`CogVideoX-5b-sat`), and training scripts seem to be built upon the **CogVideoX** project.

*   Could you please explain this discrepancy? Is Tora fundamentally an extension of CogVideoX, with OpenSora serving as a conceptual or architectural inspiration?

---

Answering this question would be helpful to determine if I can use your trajectory extractor out of the box for encoding trajectories to a compatible latent space as DiT-based Video Diffusion Models or I would have to train my own VAE from scratch. 

Any clarification you can provide on these points would be immensely helpful for the research community looking to build upon Tora's innovative ideas.

**Comment by zhenghao977** [comment]
Thank you for your interest in our work. Regarding your inquiry about the compression capabilities of the official 3D VAE trained in the RGB space, we would like to clarify that it retains the ability to compress flow maps effectively. We conducted experiments comparing our trained motion VAE used in  [Opensorav1.2](https://github.com/hpcaitech/Open-Sora/blob/main/opensora/models/vae/losses.py) with the official CogVideoX-5B-1.0 VAE. The results indicated that the CogVideoX-5B-1.0 VAE achieved approximately 2% higher motion accuracy, which we attribute to its inherently stronger performance than OpenSorav1.2.

Notably, we did not train an additional motion VAE specifically for CogVideoX-5B. However, we hypothesize that fine-tuning its existing VAE for flow visualization reconstruction could further enhance trajectory control capabilities.

**Comment by antragoudaras** [comment]
@zhenghao977 , thanks for addressing my question on the training strategy of the VAE on the RGB space ('motion-specific' fine-tuned VAE).
I have a follow-up question regarding the current codebase implementation of the TrajExtractor and the use of the 3D-VAE for encoding motion-maps to compatible DiT latents, as some points from my original issue still seem unclear.

When I trace the TrajExtractor class in traj_module.py, it appears that the visualized trajectory map (g_vis) is not passed through a 3D VAE to get a latent representation (g_m). Instead, it seems that you do something different or a simpler compression type  before the representation  gets processed by a series of ResNet-style convolutional blocks.


1. Is my understanding correct that **the current TrajExtractor implementation does not use a VAE for encoding the motion maps, bypassing the step illustrated in Figure 3 of your paper?** Could this be related to one of the ablation methods from Table 2?
2. Relating to my previous post :
> #### 3. Role of `VAETemporalEmulator` and `VAESpatialEmulator`
> 
> In the `traj_module.py` file , there are intriguing classes named [`VAETemporalEmulator`](https://github.com/alibaba/Tora/blob/main/sat/sgm/modules/traj_module.py#L132-L153) and [`VAESpatialEmulator`](https://github.com/alibaba/Tora/blob/main/sat/sgm/modules/traj_module.py#L109-L132). These seem designed to mimic parts of the VAE's functionality but are either not used or commented out in the final `TrajExtractor` implementation. After digging into the `TrajExtractor` class **it seems to me that you do not encode the visualized trajectory map (g_vis) into a latent representation (g_m) with the fine-tuned VAE you trained earlier (before TORA's training on stage 1 and stage 2 on dense and sparse trajectories respectively as described in Section 11), rather it seems that you use a simpler compression method**.
> 
> * Could you shed some light on the purpose of these emulator classes? Were they part of an earlier experimental design, and **what was the reason for not making explicit use of the 3D Motion VAE as illustrated in Figure 3 of your paper instead for encoding the motion maps to a compatible DiT latent space (before applying the ResNet convolution blocks)**?
> 
 What was these two emulator classes  intended role? How does this connects with my assumption that the VAE compression is missing in the `TrajExtractor` class.

Thanks in advance!

**Comment by Leojc** [comment]
VAETemporalEmulator and VAESpatialEmulator are part of an earlier experimental design.
